core/src/main/scala/cats/data/HashSet.scala

```diff
@@ -0,0 +1,937 @@
+        node.add(a, hash.hash(a), 0)
+      }
+
+      if (newRootNode eq rootNode)
+        this
+      else
+        new HashSet(newRootNode)
+    }
+  }
+
+  /**
+   * Typesafe equality operator.
+   *
+   * This method is similar to [[scala.Any#==]] except that it only allows two [[cats.data.HashSet]]
+   * values of the same element type to be compared to each other, and uses equality provided
+   * by [[cats.kernel.Eq]] instances, rather than using the universal equality provided by
+   * [[java.lang.Object#equals]].
+   *
+   * @param that the [[cats.data.HashSet]] to check for equality with this set.
+   * @return `true` if this set and `that` are equal, `false` otherwise.
+   */
+  def ===(that: HashSet[A]): Boolean = {
+    (this eq that) || (this.rootNode === that.rootNode)
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case set: HashSet[_] =>
+      (this eq set) || (this.rootNode == set.rootNode)
+    case _ =>
+      false
+  }
+
+  override def hashCode(): Int =
+    StaticMethods.unorderedHash(this.iterator)(this.hash)
+
+  /**
+   * Typesafe stringification operator.
+   *
+   * This method is similar to [[java.lang.Object#toString]] except that it stringifies values according
+   * to [[cats.Show]] instances.
+   *
+   * @param show the [[cats.Show]] instance to use for showing values of type `A`.
+   * @return a [[java.lang.String]] representation of this set.
+   */
+  def show(implicit show: Show[A]): String =
+    iterator.map(show.show).mkString("HashSet(", ", ", ")")
+
+  override def toString() =
+    iterator.mkString("HashSet(", ", ", ")")
+}
+
+object HashSet {
+
+  /**
+    * Creates a new empty [[cats.data.HashSet]] which uses `hash` for hashing.
+    *
+    * @param hash the [[cats.kernel.Hash]] instance used for hashing values.
+    * @return a new empty [[cats.data.HashSet]].
+    */
+  def empty[A](implicit hash: Hash[A]): HashSet[A] =
+    new HashSet(Node.empty[A])
+
+  /**
+  * Creates a new [[cats.data.HashSet]] which contains all elements of `as`.
+  *
+  * @param as the elements to add to the [[cats.data.HashSet]].
+  * @param hash the [[cats.kernel.Hash]] instance used for hashing values.
+  * @return a new [[cats.data.HashSet]] which contains all elements of `as`.
+  */
+  def apply[A](as: A*)(implicit hash: Hash[A]) =
+    fromSeq(as)
+
+  /**
+  * Creates a new [[cats.data.HashSet]] which contains all elements of `seq`.
+  *
+  * @param seq the sequence of elements to add to the [[cats.data.HashSet]].
+  * @param hash the [[cats.kernel.Hash]] instance used for hashing values.
+  * @return a new [[cats.data.HashSet]] which contains all elements of `seq`.
+  */
+  def fromSeq[A](seq: Seq[A])(implicit hash: Hash[A]): HashSet[A] = {
+    val rootNode = seq.foldLeft(Node.empty[A]) { case (node, a) =>
+      node.add(a, hash.hash(a), 0)
+    }
+    new HashSet(rootNode)
+  }
+
+  sealed abstract class Node[A] {
```

Comment:
This class probably should be private. Also its subclasses below.

Reply:
Yes good point, I haven't thought much about visibility - I left these public initially just for exploration in the REPL 😛